### PR TITLE
[AOTI XPU] Support AOT Inductor for Intel GPU.

### PR DESCRIPTION
**Pull request by etaf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140664
* __->__ #140269
* #140268

This PR add XPU support for AOT Inductor, and reuse the corresponding UT.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

**Comment by EikanWang** [comment]
Pls. update the PR description.

**Comment by etaf** [comment]
I'm splitting this PR to make it easier to review.  Convert it to draft.

**Comment by etaf** [comment]
> Pls. update the PR description.

Resolved

**Review by desertfire — COMMENTED** [review]
The same comment as the previous PR on splitting into shim_xpu.h. I will review this one after you updated the base PR.

**Review by desertfire — APPROVED** [review]

**Comment by etaf** [comment]
@EikanWang  This PR depends on #141114

File: torch/csrc/inductor/aoti_torch/shim_xpu.cpp [hunk 1]
```diff
@@ -30,8 +30,48 @@ AOTITorchError aoti_torch_xpu_guard_set_index(
       { reinterpret_cast<at::DeviceGuard*>(guard)->set_index(device_index); });
 }
 
+AOTITorchError aoti_torch_create_xpu_stream_guard(
+    void* stream,
+    int32_t device_index,
+    XPUStreamGuardHandle* ret_guard) {
+  AOTI_TORCH_CONVERT_EXCEPTION_TO_ERROR_CODE({
+    if (stream == nullptr) {
```
Comment (guangyey) [inline torch/csrc/inductor/aoti_torch/shim_xpu.cpp:38]: I quickly go through the code. I find `stream` is always `nullptr` at the beginning and initialized by `getCurrentXPUStream`, right? Then, we use `getStreamFromExternal` to convert it to `XPUStream`. If yes, this is a wrong usage scenario related to `getStreamFromExternal`.
As the [definition](https://github.com/pytorch/pytorch/blob/942a2438e263a2632b8934dd245060c9b237f4be/c10/cuda/CUDAStream.h#L219) of `getStreamFromExternal`, we can't pass an internal stream to it.

Comment (etaf) [reply torch/csrc/inductor/aoti_torch/shim_xpu.cpp]: Hi @guangyey , the stream should not be null here, the `if(stream == nullptr)` here was my workaround for a wrong test case descripted in #140805, which have I've skipped it. Now I've updated this `if(stream == null)` to `assert(stream)`. 

And in current design of AOTI , for cuda, the stream for getStreamFromExternal may be a internal or external stream. The API `AOTIModelContainerRunnerCuda::run` use current stream, and the API `AOTIModelContainerRunnerCuda::run_with_cuda_stream` use external stream.

https://github.com/pytorch/pytorch/blob/7e77c5ffba25b1a4ec3cdc60f6e31ab80ca147b2/torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp#L19C1-L32C1

**Review by EikanWang — APPROVED** [review]

**Comment by clee2000** [comment]
@pytorchbot revert -m "breaks forward compatibility?  D66937097" -c ghfirst
```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!! Forward compatibility breakage detected !!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

These symbols are declared in model so, but not in fbpkg:
aoti_torch_device_type_xpu
```

@desertfire Could you help take a look at this?

**Comment by desertfire** [comment]
> These symbols are declared in model so, but not in fbpkg:
> aoti_torch_device_type_xpu
> ```
> 
> @desertfire Could you help take a look at this?

I added a missing `#ifdef USE_XPU` in `aoti_runtime/model.h`, and tested it passes in fbcode.

**Comment by malfet** [comment]
Also, this test is breaking Mac testing, although results were initially occluded by https://github.com/pytorch/pytorch/issues/142206

**Comment by desertfire** [comment]
> Also, this test is breaking Mac testing, although results were initially occluded by #142206

Do you have a test failure link? I wonder what the problem is.

**Comment by etaf** [comment]
> > Also, this test is breaking Mac testing, although results were initially occluded by #142206
> 
> Do you have a test failure link? I wonder what the problem is.

Seems not relate to this PR: #142436

**Comment by etaf** [comment]
The failed job :[xpu / linux-jammy-xpu-2025.0-py3.9 / test (default, 4, 4, linux.idc.xpu)](https://hud.pytorch.org/pr/pytorch/pytorch/140269#34163754963) ([gh](https://github.com/pytorch/pytorch/actions/runs/12245053349/job/34163754963))
inductor/test_torchinductor_opinfo.py::TestInductorOpInfoXPU::test_comprehensive_masked_cumprod_xpu_float16

is a known issue #141861 and has been fixed in main branch by #142348.

Please ignore it.

**Pull request closed without merging**